database/DatabaseExampleSwiftUI/DatabaseExample/Shared/Models/Post.swift

```diff
@@ -47,9 +49,33 @@ struct Post: Identifiable {
     self.title = title
     self.body = body
     self.starCount = starCount
+    self.starDictionary = [:]
   }
 
   init() {
     self.init(id: "", uid: "", author: "", title: "", body: "")
   }
+
+  mutating func didTapStarButton(isStarred: Bool) {
+    let ref: DatabaseReference = {
+      Database.database().reference()
+    }()
+    if isStarred {
+      // add current user to the post.starDictionary
+      starDictionary[uid] = true
+      ref.child("/posts/\(id)/starDictionary").setValue(starDictionary)
```

Comment:
nit: looks like we are writing the entire dictionary to modify add/remove a single user. In a real application this approach will lead to data loss, because of concurrent modification attempts from different users. E.g.:
1. user1 read the dict1
2.  user2 reads the dict1
3. user1 modifies and writes dict1+user1
4. user2 modifies and writes dict1+user2

As a result we have dict1+user2 stored, but should have dict1+user1+user2. This will only get worse with the number of users.

Suggestion to mitigate: add/remove a single child representing the user ID. It still may have issues with concurrent modification by the same user from different devices, but it is more an edge case. 

Though it's not a big issue for the sample project I still would recommend fixing it eventually to provide better guidance to Firebase users.

Reply:
BTW, the same is relevant for the starCount. You may count children in the star collection instead, but it may not perform well.  You can use [increment](https://firebase.google.com/docs/reference/swift/firebasedatabase/api/reference/Classes/ServerValue?authuser=0#increment_:) method but decrementing will be more challenging to implement correctly. TBH, taking into account complexity of correct and optimal counter implementation I would suggest excluding this feature so far.

Reply:
Thank you for the feedback. I will take a closer look at updating the firebase data and use specific methods instead of updating the entire one.